### ISSUE TITLE
Validation dataset

### ISSUE THREAD

**Issue by CBvanYperen** [google-research/pegasus#131]
@registry.register("new_params")
def my_param(param_overrides):
  return public_params.transformer_params(
      {
          "train_pattern": "tfds:new_tfds_dataset,train",
          "dev_pattern": "tfds:new_tfds_dataset,validation",
          "test_pattern": "tfrecord:new_dataset_files.tfrecord*",
          "max_input_len": 512,
          "max_output_len": 128,
          "train_steps": 10000,
          "learning_rate": 0.0001,
          "batch_size": 8,
      }, param_overrides)

Is the "dev_pattern" dataset used for validation?
If so, at which moments is it used?
If so, is there a way to check the validation scores?

Your help and efforts are much appreciated!

**Comment by agenius5** [comment]
The evaluation script by default evalutes dev or validation set.

**Comment by CBvanYperen** [comment]
Thanks for your reply, so that means that running:
```
python3 pegasus/bin/evaluate.py --params=aeslc_transformer \
--param_overrides=vocab_filename=ckpt/pegasus_ckpt/c4.unigram.newline.10pct.96000.model,batch_size=1,beam_size=5,beam_alpha=0.6 \
--model_dir=ckpt/pegasus_ckpt/aeslc
```
will actually use the dev / validation set. And only if the following flag is used `--evaluate_test=True`, then it uses the test_pattern?

Also, I'm not sure exactly what continuous_eval does and where can I see the scores of continuous eval. An explanation of this would be very much appreciated!

**Comment by agenius5** [comment]
Yes, that script uses dev set unless you specify `- - evaluate_test = True`. To evaluate on full sets, you have to set another flag `--full=True`.

I ran continuous eval yesterday. What it does is train the model and evaluate dev set on intermediate checkpoints. But, it needs two gpu - one to train and one to evaluate. Scores will be saved in model directory and reported in logs too.

**Comment by CBvanYperen** [comment]
Thank you! 

So continious eval is an alternative for train, and not an alternative for evaluate? 

What part of the set does it use to evaluate if you do not set `--full=True`?

Which GPU's did you use / what would be the minimum you think?

**Comment by agenius5** [comment]
> So continious eval is an alternative for train, and not an alternative for evaluate?

Yes, it is. Basically, it gives dev set results on intermediate checkpoints while training. You can pick best checkpoint later for evaluation on test set.

> What part of the set does it use to evaluate if you do not set --full=True?

Not sure, On AESLC dataset, it used 1000 examples out of 1906. You might need to look at the code to see what part of set is evaluated.

> Which GPU's did you use / what would be the minimum you think?

I am running it on Colab. It offers Nvidia Tesla T4 most of the time and that gives train speed of 0.4 steps (epochs) per second. Sometimes I get Nvidia P100 and get speed of 0.7 step per second. This is for normal training and for continuous evaluation it needs two gpu. I am thinknig of some workound to be able to run continuous eval.

**Comment by CBvanYperen** [comment]
Thank you for the previous reply. 

I do have the option to utilize two GPU's in a GCP Notebook. If I would then want to train a model and evaluate on intermediate steps with continuous evaluation. Would I then have to run:

```
!python3 pegasus/bin/continuous_eval.py \
--params=my_parameters \
--param_overrides=vocab_filename=ckpt/pegasus_ckpt/c4.unigram.newline.10pct.96000.model \
--train_init_checkpoint=ckpt/pegasus_ckpt/model.ckpt-1500000 \
--model_dir=path_to_some_empty_directory \
```
?

Also, I could not find the --full FLAG or Hyperparameter anywhere, could you point me to where I could find it?
Thanks a lot!

**Comment by agenius5** [comment]
As far as I have understood things, there's a file name `checkpoint` inside your model output directory. So, continuous_eval looks for the latest checkpoint available from the `checkpoint` file.

And yes, I couldn't find the --FULL flag in `continuous_eval` either. I think it's only available in `evaluate.py'. I am still trying to understand `continuous_eval` script to figure how exactly it works.

**Comment by CBvanYperen** [comment]
Alright, I thought the point of continuous eval was to check for each checkpoint and not just the latest one, because I'd like to check on the validation set on each checkpoint. I guess the only way to do that is just to run the `continuous_eval` function for each available checkpoint separately.

Additionally, I was wondering whether there is any way of figuring out which checkpoint is chosen when in `evaluate.py` the `--best=True` flag is added.

**Comment by agenius5** [comment]
Suppose I train the model from 1000 step onwards till 1500 steps so now I have 5 checkpoints and latest is 1500. The checkpoint file says that 1500 is latest in the first line itself so I change it to 1100 and then run the `continuous_eval' script. It'll calculate metrics for 1100 checkpoint and then waits for next checkpoint so I change the first line in checkpoint file to 1200 and then metrics for 1200 gets calcualted.

> Additionally, I was wondering whether there is any way of figuring out which checkpoint is chosen when in evaluate.py the --best=True flag is added.

I was wondering the same, I'll try it out soon.
